AI #1803: Contract Commitment dataset extension (#1924)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/appendix/contract_commitments/contract_commitments.md

## Contract Commitments

The **Contract Commitment** dataset provides a structured representation of the commercial agreements between a customer and their service providers. While the [Cost and Usage](#datasets.costandusage) dataset tracks the results of consumption, the Contract Commitment dataset tracks the intent and constraints of the relationship.

### Core Logical Pillars

To ensure interoperability across different cloud and SaaS providers, the dataset relies on three core logical pillars:

1. **Commitment Categorization:** Distinguishes between obligations based on **Spend** (e.g., "I will spend 1M") vs. **Usage** (e.g., "I will use 500 vCPUs"). This determines which metrics—Cost or Quantity—are used to measure fulfillment.

> 2. **Fulfillment Modeling:** Defines the "reset" behavior of a benefit.

Reviewer: I left a comment lower down asking about "reset" so I see it's coming from here. Not sure what "reset behavior" is. It brings to mind for me auto-renewal, which I don't think it's intending to.

I would characterize this as the manner in which the commitment is consumed. Or the manner in which the commitment must be used by the recipient.

Author: Agree.  To be honest, this was AI slop.   I was very careful about the column definition markdowns, but this summary got it wrong.  I've repaired it not.  Resolving this comment, but feel free to reopen if you feel it is still off.

   * **Continuous** models (like Reserved Instances) are typically "use-it-or-lose-it" within short windows (e.g., Hourly).
   * **Discontinuous** models (like Enterprise Agreements) allow consumption to be aggregated over a longer duration (e.g., Total Term).
3. **Eligibility Boundaries:** Using a structured JSON format, the dataset defines the logical perimeter of a commitment, specifying exactly which accounts, regions, or services are eligible to receive the negotiated benefit.

### Expected Value Taxonomy

The following table defines the high-level expectations for key categorical columns in this dataset:

| Attribute | Expected Value Logic | Example Values |
| :--- | :--- | :--- |
| **Benefit Category** | The primary commercial advantage provided. | `Discount`, `Monetary Pool`, `Availability` |

> | **Model** | How the benefit expires or resets. | `Continuous`, `Discontinuous` |

Reviewer: I'm not tracking on why we're including "or resets." For me this brings to mind that a "Continuous" reset might mean the contract begins anew when it is over, (renews) which isn't the case. 

Is there an example of what a Continuous or Discontinuous reset is?

Reviewer: I left a better comment higher up about this topic.

Author: See my other comment.  Fixed now.

If you scroll down in this appendix, you will see several examples.  I've also added the example from [this tab](https://docs.google.com/spreadsheets/d/1XcY8wHzjbXywFlK-MDs7kQtXJ6hnWcq_ldq22csht-E/edit?gid=730749207#gid=730749207) we reviewed yesterday.  Give those a look and let me know what you think.

### Expected Value Taxonomy

The following table defines the high-level expectations for key categorical columns in this dataset:

| Attribute | Expected Value Logic | Example Values |
| :--- | :--- | :--- |
| **Benefit Category** | The primary economic advantage provided. | `Discount`, `Monetary Pool`, `Availability`, `Other` |
| **Model** | How the commitment is consumed. | `Continuous`, `Discontinuous` |
| **Fulfillment Interval** | The "Use-it-or-lose-it" or "Goal" window for reset. | `Hourly`, `Monthly`, `Annual`, `Total Term` |
| **Status** | The current lifecycle state of the record. | `Active`, `Pending`, `Expired`, `Canceled` |
| **Offer Category** | The "privacy" or source level of the pricing. | `Public`, `Negotiated` |
| **Payment Model** | The cash-flow timing for the commitment. | `No Upfront`, `Partial Upfront`, `All Upfront` |

### Why the Values Matter

By standardizing these values, organizations can move from manual spreadsheet tracking to **Automated FinOps Governance**.

> For example, a **Discontinuous** model with an **Annual Fulfillment Interval** tells a reporting engine to look for a "True-up" event at the end of the year. Conversely, a **Continuous** model with an **Hourly Fulfillment Interval** tells the engine to calculate "Waste" (unused capacity) for every individual hour of the billing period.

Reviewer 2: ```suggestion
For example, in a Discontinuous model with an Annual Fulfillment Interval, a reporting engine does not evaluate the commitment hour-by-hour. Instead, it accumulates usage across the full year and waits until the end of that annual interval to determine whether the contractual threshold has been met. At that point, it looks for a “True-up” event — the reconciliation step that settles any difference between the committed amount and actual consumption.

By contrast, in a Continuous model with an Hourly Fulfillment Interval, an engine evaluates the commitment during every hour of the billing period. For each individual hour, it compares committed capacity to actual usage. If usage falls short during a given interval, the engine calculates “Waste”, representing the unused portion of the commitment for that specific hour. That unused capacity is measured and reported as it occurs rather than deferred for end-of-term reconciliation.
```

I feel like the wording in the current example requires the reader to fully grok the terms of the math. Maybe being a little more wordy here will help the reader but also make the example something that can be read spec without losing all its context.

Author: Looks good to me!  I'll commit.

## Examples

### Common Offering Examples

The following table provides a reference for how common cloud and SaaS commercial offerings theoretically map to the Contract Commitment schema. (This table is provided for demonstration purposes only.  Actual value assignments are left to the data generators upon the creation of dataset artifacts.)

> | Offering Example | CC Category | CC Model | CC Offer Category | CC Benefit Category | CC Fulfillment Interval | CC Duration | CC Payment Model | CC Payment Interval |

Reviewer 2: Is it clear enough that CC means Contract Commitment here? I wonder if we should state this shorthand before we use it in the table.

Author: Happy to leave a simple disclaimer.  Just trying to keep the width of the table down.  What do you think of this?

```suggestion
*(In the below table, CC represents Contract Commitment.)*

| Offering Example | CC Category | CC Model | CC Offer Category | CC Benefit Category | CC Fulfillment Interval | CC Duration | CC Payment Model | CC Payment Interval |
```

# Contract Commitments

## Overview

> The **Contract Commitment** dataset provides a structured representation of the commercial agreements between a customer and their service providers. While the [Cost and Usage](#datasets.costandusage) dataset tracks the results of consumption, the Contract Commitment dataset tracks the intent and constraints of the relationship.

Reviewer 3: I propose removing "Commitment" from being a base language on this and we simply use contract.

Some contracts have commitments in them (customer agrees to a commitment of paying $AAAAAA over Y term is the most common) but not all contracts have commitments.

Author: I'm sorry to say that changing the name of the `Contract Commitment` dataset is outside the scope of this PR.  That name was approved in FOCUS 1.3, and we are merely extending the dataset here in 1.4.

I'm happy for you to author a separate PR pitching a rename of this dataset and its constituent columns -- but this is the pattern to which we shall adhere in this PR.

Resolving this comment.  Feel free to reopen it if you feel we must consider a rename as part of this PR, but please explain why, in detail.

# Contract Commitments

## Overview

The **Contract Commitment** dataset provides a structured representation of the commercial agreements between a customer and their service providers. While the [Cost and Usage](#datasets.costandusage) dataset tracks the results of consumption, the Contract Commitment dataset tracks the intent and constraints of the relationship.

### Core Logical Pillars

To ensure interoperability across different data generators, the dataset relies on three core logical pillars:

1. **Commitment Categorization:** Distinguishes between obligations based on **Spend** (e.g., "I will spend 1M") vs. **Usage** (e.g., "I will use 500 vCPUs"). This determines which metrics —- Cost or Quantity —- are used to measure fulfillment.
2. **Fulfillment Modeling:** Defines the operational behavior and consumption flexibility of a commitment.

>    * **Continuous** models (like Reserved Instances) are typically "use-it-or-lose-it" within short windows (e.g., Hourly).

Reviewer 4: ```suggestion
   * **Continuous** models (like Reserved Instances) are "use-it-or-lose-it", typically within short windows (e.g., Hourly).
```

   * **Discontinuous** models (like Enterprise Agreements) allow consumption to be aggregated over a longer duration (e.g., Total Term).
3. **Eligibility Boundaries:** Using a structured JSON format, the dataset defines the logical perimeter of a commitment, specifying exactly which accounts, regions, or services are eligible to receive the negotiated benefit.

### Expected Value Taxonomy

The following table defines the high-level expectations for key categorical columns in this dataset:

| Attribute | Expected Value Logic | Example Values |
| :--- | :--- | :--- |
| **Benefit Category** | The primary economic advantage provided. | `Discount`, `Monetary Pool`, `Availability`, `Other` |
| **Model** | How the commitment is consumed. | `Continuous`, `Discontinuous` |
| **Fulfillment Interval** | The "Use-it-or-lose-it" or "Goal" window for reset. | `Hourly`, `Monthly`, `Annual`, `Total Term` |

> | **Status** | The current lifecycle state of the record. | `Active`, `Pending`, `Expired`, `Canceled` |

Author: Note to self: update the fields with the slight renames we have sustained via TF-1 review.

Author: This was accomplished via @Matt-Cowsert's suggestions.  Resolving.

By contrast, in a Continuous model with an Hourly Fulfillment Interval, an engine evaluates the commitment during every hour of the billing period. For each individual hour, it compares committed capacity to actual usage. If usage falls short during a given interval, the engine calculates "Waste", representing the unused portion of the commitment for that specific hour. That unused capacity is measured and reported as it occurs rather than deferred for end-of-term reconciliation.

## Examples

### Common Offering Examples

The following table provides a reference for how common cloud and SaaS commercial offerings theoretically map to the Contract Commitment schema. (This table is provided for demonstration purposes only.  Actual value assignments are left to the data generators upon the creation of dataset artifacts.)

In the below table, CC represents Contract Commitment.

> | Offering Example | CC Category | CC Model | CC Offer Category | CC Benefit Category | CC Fulfillment Interval | CC Duration | CC Payment Model | CC Payment Interval |

Reviewer 5: ```suggestion
| Offering Example | CC Category | CC Model | CC Offer Category | CC Benefit Category | CC Fulfillment Interval | CC Duration Type | CC Payment Model | CC Payment Interval |
```

Column was renamed from ContractCommitmentDuration to ContractCommitmentDurationType. Table header should reflect the current name.

#### Commitment 3: Marketplace SaaS Add-on

* **Context:** A specialized analytics tool, **DataStreamer**, purchased through the Acme marketplace.
* **Commercial Logic:** A **Spend-based** "pass-through" for financial tracking with an **Annual Fulfillment Interval**.
* **The Issuer/Provider Split:** The **Service Provider** is **DataStreamer**, but the **Invoice Issuer** remains **Acme Co**, showing how the model tracks third-party spend in a unified ecosystem.

### Data Example: AGR-99-BETA

| Column | Commitment 1: Spend Pool | Commitment 2: Compute RI | Commitment 3: Marketplace SaaS |
| :--- | :--- | :--- | :--- |
| **Billing Currency** | `EUR` | `EUR` | `EUR` |
| **CC Benefit Category** | `Monetary Pool` | `Discount` | `Monetary Pool` |
| **CC Category** | `Spend` | `Usage` | `Spend` |
| **CC Cost** | `925000.00` | `46250.00` | `111000.00` |
| **CC Created** | `2025-12-01T09:00:00Z` | `2025-12-01T09:00:00Z` | `2026-01-15T14:30:00Z` |
| **CC Description** | `3yr Enterprise Spend Goal` | `us-east-1 m5 Reservations` | `DataStreamer Pro via Marketplace` |
| **CC Discount %** | `0.15` | `0.40` | `0.10` |

> | **CC Duration** | `3 Years` | `1 Year` | `1 Year` |

Reviewer 5: ```suggestion
| **CC Duration Type** | `3 Years` | `1 Year` | `1 Year` |
```

Renamed column: ContractCommitmentDurationType.

> | **CC Eligibility** | `{"IsGlobalScope": true}` | `{"Inclusions": [{"Dimension": "RegionId", "Operator": "In", "Values": ["us-east-1"]}]}` | `{"Inclusions": [{"Dimension": "ServiceCategory", "Operator": "In", "Values": ["Analytics"]}]}` |

Reviewer 5: ```suggestion
| **CC Applicability** | `{"IsGlobalScope": true}` | `{"InclusionOperator": "Or", "Inclusions": [{"Dimension": "RegionId", "Operator": "In", "Values": ["us-east-1"]}]}` | `{"InclusionOperator": "Or", "Inclusions": [{"Dimension": "ServiceCategory", "Operator": "In", "Values": ["Analytics"]}]}` |
```

Two fixes: (1) Column was renamed from Eligibility to Applicability. (2) Non-GlobalScope JSON objects are missing `InclusionOperator`, which the ContractCommitmentApplicabilityObjectSchema in this PR requires when neither `IsGlobalScope` nor `IsComplexScope` is true.

#### Commitment 2: Security Seat License (Quantity-based)

* **Context:** A commitment to **500 Seats** of an endpoint security platform.
* **Commercial Logic:** A **Quantity-based**, **Discontinuous** model. The unit of measure is **Seats** rather than a currency value.
* **Invoice/Provider Alignment:** Unlike the Marketplace example, this is billed directly by the vendor (**CyberGuard Inc**), yet remains logically associated with the broader Cloud Transformation contract.

#### Commitment 3: Cross-Cloud Data Connector (Tiered Usage)

* **Context:** A commitment based on **Data Volume (TB)** specifically for egress traffic between cloud providers.
* **Commercial Logic:** A **Usage-based**, **Continuous** model tracked on a **Monthly Fulfillment Interval**.
* **Eligibility:** Targeted specifically at `Egress` usage types via string-match logic in the Eligibility JSON.

### Data Example: AGR-44-GAMMA

| Column | Commitment 1: AI Training | Commitment 2: Security Seats | Commitment 3: Data Connector |
| :--- | :--- | :--- | :--- |
| **Billing Currency** | `USD` | `USD` | `USD` |
| **CC Benefit Category** | `Discount` | `Availability` | `Discount` |

> | **CC Category** | `Usage` | `Quantity` | `Usage` |

Reviewer 5: ```suggestion
| **CC Category** | `Usage` | `Usage` | `Usage` |
```

"Quantity" isn't currently a valid ContractCommitmentCategory value. The allowed values (per 1.3, unchanged in this PR) are "Spend" and "Usage". I'd recommend "Usage" here since a seat-based commitment represents a predetermined amount of usage measured in Seats. If you'd prefer "Quantity" as a distinct third category, contractcommitmentcategory.md would need updating in this PR.

Author: Good call.  Committing.

| **CC Cost** | `250000.00` | `120000.00` | `15000.00` |
| **CC Created** | `2026-02-01T08:00:00Z` | `2026-02-01T08:00:00Z` | `2026-02-01T08:00:00Z` |
| **CC Description** | `H100 GPU Reservation - Q1` | `CyberGuard Endpoint Seats` | `Inter-Cloud Egress Tier` |
| **CC Discount %** | `0.30` | `null` | `0.50` |

> | **CC Duration** | `3 Months` | `1 Year` | `1 Year` |

Reviewer 5: ```suggestion
| **CC Duration Type** | `3 Months` | `1 Year` | `1 Year` |
```

Renamed column: ContractCommitmentDurationType.

> | **CC Eligibility** | `{"Inclusions": [{"Dimension": "ServiceCategory", "Operator": "In", "Values": ["AI/ML"]}]}` | `{"IsGlobalScope": true}` | `{"Inclusions": [{"Dimension": "UsageType", "Operator": "Contains", "Values": ["Egress"]}]}` |

Reviewer 5: ```suggestion
| **CC Applicability** | `{"InclusionOperator": "Or", "Inclusions": [{"Dimension": "ServiceCategory", "Operator": "In", "Values": ["AI/ML"]}]}` | `{"IsGlobalScope": true}` | `{"InclusionOperator": "Or", "Inclusions": [{"Dimension": "UsageType", "Operator": "Contains", "Values": ["Egress"]}]}` |
```

Same two fixes: Eligibility → Applicability rename, and missing `InclusionOperator` in non-GlobalScope JSON objects.

#### Commitment 3: CDN Annual (Volume Exhaustion)

* **Context:** An annual volume commitment of 1PB (1,000TB) for Content Delivery Network services.
* **Commercial Logic:** This is a **Discontinuous** model with an **Annual Fulfillment Interval**.
* **Overage Status:** Because the customer has already consumed their allotted volume before the `CC Period End`, the **CC Status** has shifted to `Overage`. This signals that the pool is empty and subsequent usage will be handled according to the contract's true-up or on-demand terms.

### Data Example: AGR-11-DELTA

| Column | Commitment 1: Base Storage | Commitment 2: Storage Overage | Commitment 3: CDN Annual |
| :--- | :--- | :--- | :--- |
| **Billing Currency** | `USD` | `USD` | `USD` |
| **CC Benefit Category** | `Discount` | `Discount` | `Monetary Pool` |
| **CC Category** | `Usage` | `Usage` | `Usage` |
| **CC Cost** | `50000.00` | `0.00` | `100000.00` |
| **CC Created** | `2026-02-01T08:00:00Z` | `2026-02-01T08:00:00Z` | `2026-02-01T08:00:00Z` |
| **CC Description** | `Base 100TB DB Storage` | `Tier 2 Storage Overage` | `1PB Annual CDN Volume` |
| **CC Discount %** | `0.20` | `0.10` | `0.25` |

> | **CC Duration** | `1 Year` | `1 Year` | `1 Year` |

Reviewer 5: ```suggestion
| **CC Duration Type** | `1 Year` | `1 Year` | `1 Year` |
```

Renamed column: ContractCommitmentDurationType.

> | **CC Eligibility** | `{"Inclusions": [{"Dimension": "ServiceCategory", "Operator": "In", "Values": ["Database"]}]}` | `{"Inclusions": [{"Dimension": "ServiceCategory", "Operator": "In", "Values": ["Database"]}]}` | `{"Inclusions": [{"Dimension": "ServiceCategory", "Operator": "In", "Values": ["CDN"]}]}` |

Reviewer 5: ```suggestion
| **CC Applicability** | `{"InclusionOperator": "Or", "Inclusions": [{"Dimension": "ServiceCategory", "Operator": "In", "Values": ["Database"]}]}` | `{"InclusionOperator": "Or", "Inclusions": [{"Dimension": "ServiceCategory", "Operator": "In", "Values": ["Database"]}]}` | `{"InclusionOperator": "Or", "Inclusions": [{"Dimension": "ServiceCategory", "Operator": "In", "Values": ["CDN"]}]}` |
```

Same two fixes: Eligibility → Applicability rename, and missing `InclusionOperator`.

| **CC Fulfillment Interval** | `Monthly` | `Monthly` | `Annual` |
| **CC ID** | `CMT-STR-BASE` | `CMT-STR-OVER` | `CMT-CDN-VOL` |
| **CC Last Updated** | `2026-02-01T08:00:00Z` | `2026-02-01T08:00:00Z` | `2026-02-01T08:00:00Z` |
| **CC Model** | `Continuous` | `Continuous` | `Discontinuous` |
| **CC Offer Category** | `Negotiated` | `Negotiated` | `Negotiated` |
| **CC Payment Interval** | `Monthly` | `Monthly` | `Annual` |
| **CC Payment Model** | `No Upfront` | `No Upfront` | `All Upfront` |
| **CC Payment Upfront %** | `0.00` | `0.00` | `1.00` |
| **CC Period End** | `2027-02-01` | `2027-02-01` | `2027-02-01` |
| **CC Period Start** | `2026-02-01` | `2026-02-01` | `2026-02-01` |
| **CC Quantity** | `100.00` | `0.00` | `1000.00` |

> | **CC Status** | `Active` | `Active` | `Overage` |

Reviewer 5: "Overage" isn't currently in the ContractCommitmentLifecycleStatus allowed values (Proposed, Pending, Active, Expired, Canceled, Superseded). The narrative here describes a distinct state worth capturing: pool exhausted, still within term. I'd recommend adding "Overage" as a new allowed value in contractcommitmentlifecyclestatus.md (e.g., Sort Order 35, between Active and Expired) rather than forcing this into an existing value.

Author: Interesting.  Let me think about this one.

Author: @Matt-Cowsert 
Upon reflection, I think this is a very good call.  I decided to go with `Exhausted` to be more in-line with the adjective nature of the other values.  Let me know what you think and if we can resolve!

Reviewer 5: I like it @shawnalpay, please resolve at your convenience.

## Examples

### Common Offering Examples

The following table provides a reference for how common cloud and SaaS commercial offerings theoretically map to the Contract Commitment schema. (This table is provided for demonstration purposes only.  Actual value assignments are left to the data generators upon the creation of dataset artifacts.)

In the below table, CC represents Contract Commitment.

| Offering Example | CC Category | CC Model | CC Offer Category | CC Benefit Category | CC Fulfillment Interval | CC Duration Type | CC Payment Model | CC Payment Interval |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| **AWS Savings Plan** | Spend | Continuous | Public | Discount | Hourly | 1 Year | Partial Upfront | Monthly |
| **Azure Reservation** | Usage | Continuous | Public | Discount | Hourly | 3 Years | All Upfront | One-Time |
| **GCP Flex CUD** | Spend | Continuous | Public | Discount | Hourly | 3 Years | No Upfront | Monthly |
| **Azure ODCR** | Usage | Continuous | Public | Availability | Hourly | 1 Month | No Upfront | Monthly |

> | **AWS EDP / MACC** | Spend | Discontinuous | Negotiated | Discount | Annual | 3 Years | No Upfront | Monthly |

Reviewer 4: Should EDP and MACC be reclassified as `Continuous` with an `Annual` Fulfillment Interval rather than `Discontinuous`?

* They represent a flat, constant spend floor
* The discount rate is fixed per Fulfillment Interval (Annual)
* Benefits are not carried over — if the committed spend is not met, the shortfall is charged

Attaching the `Continuous` allowed value definition for context:

> *A flat, constant 'floor' of commitment. Coverage is applied at a fixed rate per Contract Commitment Fulfillment Interval (usually hourly), and benefits are not carried over to subsequent intervals.*

```suggestion
| **AWS EDP / MACC** | Spend | Continuous| Negotiated | Discount | Annual | 3 Years | No Upfront | Monthly |
```

Author: Let's see what @robmartin33 thinks.

Reviewer: If you have a 3 year one of these, do you need to spend a certain amount each year? Or do you simply need to spend all of it by the end of year 3? I think it's the latter, but things may have changed since I did contracts. 

Keep in mind I think you can also have a 1 year one of these. Or a 6 year one. Again, unless things have really changed. 

I would argue that this is discontinuous because you are not required to spend on a "fixed rate per Contract Commitment Fulfillment interval" 

In my experience, there was a grace period for spending in year 1 that allowed it to carry over to year 2, etc. and even beyond the scope of the EDP there was a grace period. from a technically detailed perspective, this might be described as you say, but from a realistic perspective I think most people would perceive this as "fulfilling" monthly when the invoice comes, and any annual requirement would be secondary. 

I don't know how precise we need to ever get with these descriptions. But to me Discontinuous captures the sentiment of an EDP in that I can spend on it AS MUCH or AS LITTLE per billing period as I like, so long as I have used it all up at the end (or got charged for it. )

Author: Thank you for that @robmartin33!

Discussed with @ijurica.  Agree that if the Fulfillment Interval and the Duration are the same, then that's Discontinuous, and we should add examples for both here.  I will do that.

Author: @ijurica I have now addressed this via a separate commit -- both the definition for CC Model, and the appendix examples.  Can you give it a look over and resolve this comment if you agree?

Author: Discussed and agreed in Mar 10 TF-1, can resolve.

* **Context:** A high-level Enterprise Agreement (EA) where the customer commits to spending **1M USD** over three years.
* **Commercial Logic:** A **Spend-based**, **Discontinuous** model. Every dollar spent within the three-year window fulfills the commitment.
* **Eligibility:** **Global**. Applies to any service or region.

#### Commitment 2: Regional Compute Reservations

* **Context:** Fixed capacity of virtual machines in a specific data center for stable production workloads.
* **Commercial Logic:** A **Usage-based**, **Continuous** model. This benefit is "use it or lose it" on an **Hourly** basis.
* **Eligibility:** Restricted to specific resource types running in the `us-east-1` region.

#### Commitment 3: Marketplace SaaS Add-on

* **Context:** A specialized analytics tool, **DataStreamer**, purchased through the Acme marketplace.
* **Commercial Logic:** A **Spend-based** "pass-through" for financial tracking with an **Annual Fulfillment Interval**.
* **The Issuer/Provider Split:** The **Service Provider** is **DataStreamer**, but the **Invoice Issuer** remains **Acme Co**, showing how the model tracks third-party spend in a unified ecosystem.

> ### Data Example: AGR-99-BETA

Reviewer 6: I'm not fully confident where this should go.

but I want to think through the implementation of partial up-front commitment where there may be differing discounts for the up-front portion vs the monthly reoccurring portion.

Example:  A 1-year Savings Plan, 50% up-front, up-front discount being 15%, and monthly reoccurring discount being 10%.

The singular effective discount would be 12.5%, however, I believe this would require 2 entries to properly represent this.

All of the columns would have the same value EXCEPT.
Description (likely should give a description of which portion they each represent, but not required to be same or different)
CC Discount (.15 vs .1)
CC Payment Interval (One Time vs Monthly)
CC Payment model (I actually think this would be "Partial Upfront" for both)
CC Payment Upfront %: (1 vs 0)

However, would it be acceptable to have multiple entries for a single contract commitment with the same CC ID, and both set to Active?

Author: @benihana88 
I'm struggling to grok this part of your example:

> Example: A 1-year Savings Plan, 50% up-front, up-front discount being 15%, and monthly reoccurring discount being 10%.

Up-front discount being 15% of... what?  Monthly reoccurring discount being 10% of... what?

An SP or RI will yield a particular percentage discount of a unit price, based on the upfront-ness of your payment model.  Let's say All Upfront yields a 15% discount; No Upfront yields 10%, and Partial Upfront yields 12.5%.  Each of these would be separate agreements.

In your example, it would be a single entry of Partial Upfront, with the Payment Upfront % being 50%, and the Discount % being whatever the agreed/marketed discount percentage, which in your example is 12.5%.  If you wanted to state 10% or 15% as separate rows, I go back to my original question of... a discount of _what_, precisely?

I propose no action on this PR.  Happy to discuss with you 1:1 further if I've misunderstood your example/point!

Reviewer 6: I think this is a side case that isn't fully fleshed out, and shouldn't stop the overall process.  No concerns about setting it aside for now and addressing in the future if others find similar use cases.
